### PR TITLE
[Feature] Add feature refine CUDA op for rotated detection. 

### PR DESCRIPTION
**Pull request by zytx121**
## Motivation

Add feature refine CUDA operation for rotate detection. 

```
@inproceedings{yang2021r3det,
    title={R3Det: Refined Single-Stage Detector with Feature Refinement for Rotating Object},
    author={Yang, Xue and Yan, Junchi and Feng, Ziming and He, Tao},
    booktitle={Proceedings of the AAAI Conference on Artificial Intelligence},
    volume={35},
    number={4},
    pages={3163--3171},
    year={2021}
}

```

## Modification

Several files in mmcv/ops folder.

## BC-breaking (Optional)

No.

## Use cases (Optional)

from mmcv.ops import feature_refine

### CODE REVIEW

File: mmcv/ops/csrc/common/cuda/feature_refine_cuda_kernel.cuh [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+#define FEATURE_REFINE_CUDA_KERNEL_CUH
+
+#ifdef MMCV_USE_PARROTS
+#include "parrots_cuda_helper.hpp"
+#else
+#include "pytorch_cuda_helper.hpp"
+#endif
+
+template <typename scalar_t>
+__global__ void feature_refine_forward_kernel(
```
Comment (zhouzaida) [inline mmcv/ops/csrc/common/cuda/feature_refine_cuda_kernel.cuh:14]:
```suggestion
__global__ void rotated_feature_align_forward_kernel(
```

Comment (zhouzaida) [reply mmcv/ops/csrc/common/cuda/feature_refine_cuda_kernel.cuh:14]: Should we need to replace all the `feature_refine` with `rotated_feature_align`?

File: mmcv/ops/csrc/pytorch/cuda/cudabind.cpp [hunk 2]
```diff
@@ -1358,7 +1358,49 @@ void dynamic_voxelize_forward_impl(const at::Tensor& points, at::Tensor& coors,
                                    const std::vector<float> voxel_size,
                                    const std::vector<float> coors_range,
                                    const int NDim);
+
 REGISTER_DEVICE_IMPL(hard_voxelize_forward_impl, CUDA,
                      hard_voxelize_forward_cuda);
 REGISTER_DEVICE_IMPL(dynamic_voxelize_forward_impl, CUDA,
                      dynamic_voxelize_forward_cuda);
+
+void FRForwardLauncher(const Tensor features, const Tensor best_bboxes,
```
Comment (zhouzaida) [inline mmcv/ops/csrc/pytorch/cuda/cudabind.cpp:1367]:
```suggestion
void RotatedFeatureAlignForwardLauncher(const Tensor features, const Tensor best_bboxes,
```

Comment (zhouzaida) [reply mmcv/ops/csrc/pytorch/cuda/cudabind.cpp:1367]: Should we need to replace all the `FR` with `RotatedFeatureAlign`?

File: mmcv/ops/rotated_feature_align.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        """
+        Args:
+            features (torch.Tensor): Input features with shape [N,C,H,W].
+            best_rbboxes (torch.Tensor): Refined rotate anchors with
+                shape [N,H,W,5]. Coordinate format (cx,cx,h,w,a).
+            spatial_scale (float): The scale of feature map size and
+                input image size.
+            points (int, optional): The number of sample points.
+                Only 1 and 5 are supported. Defaults to 1.
+        Returns:
```
Comment (zhouzaida) [inline mmcv/ops/rotated_feature_align.py:33]:
```suggestion
                Only 1 and 5 are supported. Defaults to 1.

        Returns:
```

File: mmcv/ops/rotated_feature_align.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        return output
+
+    @staticmethod
+    @once_differentiable
+    def backward(ctx, grad_output):
+        """
+        Args:
+            grad_output (torch.Tensor): The gradiant of output features
+                with shape [N,C,H,W].
+        Returns:
```
Comment (zhouzaida) [inline mmcv/ops/rotated_feature_align.py:52]:
```suggestion
                with shape [N,C,H,W].
                
        Returns:
```

File: mmcv/ops/csrc/common/cuda/feature_refine_cuda_kernel.cuh (hunk 1 above)
Comment (zhouzaida) [reply mmcv/ops/csrc/common/cuda/feature_refine_cuda_kernel.cuh:14]: Note that we also need to rename the filename.

File: mmcv/ops/csrc/common/cuda/feature_refine_cuda_kernel.cuh [hunk 5]
```diff
@@ -0,0 +1,129 @@
+// Copyright (c) OpenMMLab. All rights reserved.
+// Modified from
+// https://github.com/SJTU-Thinklab-Det/r3det-on-mmdetection/blob/master/mmdet/ops/fr/src/feature_refine_kernel.cu
+#ifndef FEATURE_REFINE_CUDA_KERNEL_CUH
```
Comment (zhouzaida) [inline mmcv/ops/csrc/common/cuda/feature_refine_cuda_kernel.cuh:4]: rename

File: mmcv/ops/csrc/pytorch/cuda/rotated_feature_align_cuda.cu [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+void RotatedFeatureAlignForwardLauncher(const Tensor features,
+                                        const Tensor best_bboxes,
+                                        const float spatial_scale,
+                                        const int points, Tensor output) {
+  at::cuda::CUDAGuard device_guard(features.device());
+  cudaStream_t stream = at::cuda::getCurrentCUDAStream();
+  const int output_size = features.numel();
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(
+      features.scalar_type(), "RotatedFeatureAlignForwardLaucherFun", ([&] {
```
Comment (zhouzaida) [inline mmcv/ops/csrc/pytorch/cuda/rotated_feature_align_cuda.cu:15]:
```suggestion
      features.scalar_type(), "rotated_feature_align_forward_cuda_kernel", ([&] {
```

File: mmcv/ops/csrc/pytorch/cuda/rotated_feature_align_cuda.cu [hunk 7]
```diff
@@ -0,0 +1,49 @@
+// Copyright (c) OpenMMLab. All rights reserved.
+// Modified from
+// https://github.com/SJTU-Thinklab-Det/r3det-on-mmdetection/blob/master/mmdet/ops/fr/src/feature_refine_kernel.cu
+#include "pytorch_cuda_helper.hpp"
+#include "rotated_feature_align_cuda_kernel.cuh"
+
+void RotatedFeatureAlignForwardLauncher(const Tensor features,
```
Comment (zhouzaida) [inline mmcv/ops/csrc/pytorch/cuda/rotated_feature_align_cuda.cu:7]:
```suggestion
void RotatedFeatureAlignForwardCUDAKernelLauncher(const Tensor features,
```

File: mmcv/ops/csrc/pytorch/cuda/rotated_feature_align_cuda.cu [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+        rotated_feature_align_forward_kernel<scalar_t>
+            <<<GET_BLOCKS(output_size), THREADS_PER_BLOCK, 0, stream>>>(
+                output_size, points, bottom_data, bboxes_data,
+                scalar_t(spatial_scale), features.size(1), features.size(2),
+                features.size(3), top_data);
+      }));
+  AT_CUDA_CHECK(cudaGetLastError());
+}
+
+void RotatedFeatureAlignBackwardLauncher(const Tensor top_grad,
```
Comment (zhouzaida) [inline mmcv/ops/csrc/pytorch/cuda/rotated_feature_align_cuda.cu:29]:
```suggestion
void RotatedFeatureAlignBackwardCUDAKernelLauncher(const Tensor top_grad,
```

File: mmcv/ops/csrc/pytorch/cuda/rotated_feature_align_cuda.cu [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+                                         const Tensor best_bboxes,
+                                         const float spatial_scale,
+                                         const int points, Tensor bottom_grad) {
+  at::cuda::CUDAGuard device_guard(top_grad.device());
+  cudaStream_t stream = at::cuda::getCurrentCUDAStream();
+  const int output_size = top_grad.numel();
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(
+      top_grad.scalar_type(), "RotatedFeatureAlignBackwardLaucherFun", ([&] {
```
Comment (zhouzaida) [inline mmcv/ops/csrc/pytorch/cuda/rotated_feature_align_cuda.cu:37]:
```suggestion
      top_grad.scalar_type(), "rotated_feature_align_backward_cuda_kernel", ([&] {
```

**Comment by zhouzaida** [comment]
Unit test failed in my local machine.
image
More details can be found at https://github.com/pytorch/pytorch/issues/16940.

**Comment by zytx121** [comment]
> Unit test failed in my local machine. image More details can be found at [pytorch/pytorch#16940](https://github.com/pytorch/pytorch/issues/16940).

Thanks for your web link. We have fixed it.

**Review by zhouzaida — APPROVED** [review]

**Review by ZwwWayne — APPROVED** [review]

**Pull request merged**